Bug 1435099: Audit Log events do not report default DB

gl-sergei opened the pull request:
Blueprint
https://blueprints.launchpad.net/percona-server/+spec/audit-log-default-db

Add field `DB` for `MYSQL_AUDIT_GENERAL_STATUS` records.

gl-sergei commented:
http://jenkins.percona.com/view/PS%205.6/job/percona-server-5.6-param/1227/

laurynas-biveinis commented on mysql-test/t/audit_log_default_db.test:
```diff
@@ -0,0 +1,74 @@
+connection default;
+
+--exec $MYSQL --user=user1 --password=111 test -e "use db1; SELECT * FROM t;"
+# is there better way to check that server reset connection initiated by mysql cli?
```
Wrapping exec with include/count_connections and include/wait_until_connection_count should do this

laurynas-biveinis commented on mysql-test/t/audit_log_echo.inc:
```diff
@@ -0,0 +1,26 @@
+# Echo the contents of hte audit log (in CSV format)
```
s/hte/the

laurynas-biveinis commented on plugin/audit_log/audit_log.c:
```diff
@@ -785,6 +796,39 @@ int is_event_class_allowed_by_policy(unsigned int class,
   return (class_mask[policy] & (1 << class)) != 0;
 }
 
+static const CHARSET_INFO *charset_info= &my_charset_latin1;
```
Is the charset of strings passed to plugin callbacks specified anywhere?

laurynas-biveinis commented on plugin/audit_log/audit_log.c:
```diff
@@ -785,6 +796,39 @@ int is_event_class_allowed_by_policy(unsigned int class,
+  }
+
+  *len= 0;
+  while (str[*len] && my_isvar(charset_info, str[*len]))
```
Specifically, for this? Would 
use 'ąžąžąžą' 
work?

gl-sergei replied:
this patch works OK, here is what I see in the log:

```
"Query","<ID>","<DATETIME>","change_db","<CONN_ID>",0,"use `ąžąžąžą`","user1[user1] @ localhost []","localhost","","","ąžąžąžą"
```

I can't `use ąžąžąžą`, I need to `use `ąžąžąžą``

What doesn't seem to work is disconnect event... and maybe connect as well. Looks like MySQL passing database wrong.

```
"Quit","<ID>","<DATETIME>","<CONN_ID>",0,"user1","user1","","","localhost","","ä…å¾ä…å¾ä…å¾ä…"
```

In this line database printed from the `mysql_event_connection`. Why is this happening?

gl-sergei replied:
Hmm... it is even different on Linux and OSX!

```
-"Quit","<ID>","<DATETIME>","<CONN_ID>",0,"user1","user1","","","localhost","","ä…å¾ä…å¾ä…å¾ä…"
+"Quit","<ID>","<DATETIME>","<CONN_ID>",0,"user1","user1","","","localhost","","Ä…Å¾Ä…Å¾Ä…Å¾Ä…"
```

laurynas-biveinis replied:
Probably raw UTF-8 bytes? The Linux and OSX difference probably has to do with case-sensitive vs case-insensitive filesystem

gl-sergei replied:
I found what was wrong! `mtr` communicating with server in `latin1` by default. It was the source of these funny issues.
When I `set names utf8` I can `use ąžąžąžą` without backticks. And it was an issue for plugin. Solution is to use `event_general->general_charset` when parsing the query.
Database name in the connection event comes in `utf8mb3`.
So, there is another bug! Query is written to log in client charset while other fields in `utf8`.

laurynas-biveinis commented:
No testcase for the install plugin with an open parallel connection?

gl-sergei commented:
-   fixed typo
-   changed test to use utf8
-   added test for database 'ąžąžąžą'
-   surrounded `--exec mysql` with `--source include/count_sessions.inc/--source include/wait_until_count_sessions.inc`
-   added test for `INSTALL PLUGIN`

jenkins build: http://jenkins.percona.com/view/PS%205.6/job/percona-server-5.6-param/1231/

laurynas-biveinis commented on mysql-test/t/audit_log_default_db.test:
```diff
@@ -0,0 +1,96 @@
+GRANT ALL PRIVILEGES ON ąžąžąžą.* TO 'user2'@'%';
+
+# truncate audit log
+UNINSTALL PLUGIN audit_log;
```
This gives "plugin is busy and will be uninstalled on shutdown"

laurynas-biveinis commented on mysql-test/t/audit_log_default_db.test:
```diff
@@ -0,0 +1,96 @@
+
+connect (root,localhost,root,,,);
+connection root;
+INSTALL PLUGIN audit_log SONAME 'audit_log.so';
```
...thus this is probably a no-op WRT plugin loading and testing existing connections?

gl-sergei replied:
I disagree about INSTALL/UNINSTALL. Here is what you see in the log:

```
"Query","<ID>","<DATETIME>","install_plugin","<CONN_ID>",0,"INSTALL PLUGIN audit_log SONAME 'audit_log.so'","root[root] @ localhost []","localhost","","",""
"Quit","<ID>","<DATETIME>","<CONN_ID>",0,"root","root","","","localhost","","test"
"Query","<ID>","<DATETIME>","select","<CONN_ID>",0,"SELECT * FROM t","user1[user1] @ localhost []","localhost","","",""
"Query","<ID>","<DATETIME>","change_db","<CONN_ID>",0,"use  `db1`","user1[user1] @ localhost []","localhost","","","db1"
```

 Specifically, the query `SELECT * FROM t` doesn't report default database, but connection `test` connected with database `db1`. I also see the `Audit` record indicating that plugin is loaded. I had to grep it out, because it contains server startup options.

laurynas-biveinis replied:
OK

laurynas-biveinis commented on plugin/audit_log/audit_log.c:
```diff
@@ -785,6 +796,38 @@ int is_event_class_allowed_by_policy(unsigned int class,
 
+static
+const char *next_word(const char *str, size_t *len,
+                      struct charset_info_st *charset)
```
const arg?

laurynas-biveinis commented:
Something else about encodings, in e.g. XML the header is <?xml version=\"1.0\" encoding=\"UTF-8\"?>. If the server encoding is different, are the strings transcoded before logging?

gl-sergei commented:
> Something else about encodings, in e.g. XML the header is <?xml version=\"1.0\" encoding=\"UTF-8\"?>. If the server encoding is different, are the strings transcoded before logging?

See my comment above:

> So, there is another bug! Query is written to log in client charset while other fields in utf8.

laurynas-biveinis commented:
Ah, OK, it's the same thing we are talking about :) Have you logged the bug?

gl-sergei commented:
- `s/struct charset_info_st *charset/const struct charset_info_st *charset/`
- filed bug https://bugs.launchpad.net/percona-server/+bug/1602986

laurynas-biveinis commented:
LGTM
